File: 875_koko_eating_bananas/solution.ts
// Brute force
function minEatingSpeed(piles: number[], h: number): number {
  let k = 1
  let max = Math.max(...piles)
  let res = max

  while (k <= max) {
    let count = 0
    for (const pile of piles) {
      const c = Math.ceil(pile / k)
      count += c
    }
    if (count <= h) {
      res = Math.min(res, k)
    }
    k += 1
  }
  return piles.length + 1
}

// Optimized with binary search
function minEatingSpeed(piles: number[], h: number): number {
  let left = 1
  let right = Math.max(...piles)
  let res = right

  while (left <= right) {
    const k = Math.floor(left + (right - left) / 2)
    let hours = 0

    for (const pile of piles) {
      hours += Math.ceil(pile / k)
    }

    if (hours <= h) {
      res = Math.min(res, k)
      right = k - 1
    } else {
      left = k + 1
    }
  }

  return res
}

File: 875_koko_eating_bananas/solution.py
from math import ceil
from typing import List


class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        max_value = max(piles)
        res = max_value

        left, right = 1, max_value

        while left <= right:
            mid = left + (right - left) // 2
            hours = 0

            for pile in piles:
                hours += ceil(pile / mid)

            if hours <= h:
                res = min(res, mid)
                right = mid - 1
            else:
                left = mid + 1

        return res


print(Solution().minEatingSpeed(piles=[3, 6, 7, 11], h=8))  # 4
print(Solution().minEatingSpeed(piles=[30, 11, 23, 4, 20], h=5))  # 30
print(Solution().minEatingSpeed(piles=[30, 11, 23, 4, 20], h=6))  # 23
print(Solution().minEatingSpeed(piles=[312884470], h=312884469))  # 2
